### PR TITLE
.Net: Implementing cosine similarity in duckdb engine

### PR DESCRIPTION
**Pull request by colombod**
### Motivation and Context

This pr moves cosine similarity calculation into duckDb engine to improve performance and memory usage.

### Description

Moving cosine similarity calculation on the server there is no need to marshal out more data than the one scoring enough and meeting the result limit.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/DuckDBMemoryStore.cs [hunk 1]
```diff
@@ -152,22 +150,16 @@ public async Task RemoveBatchAsync(string collectionName, IEnumerable<string> ke
             yield break;
         }
 
-        var collectionMemories = new List<MemoryRecord>();
         TopNCollection<MemoryRecord> embeddings = new(limit);
 
-        await foreach (var record in this.GetAllAsync(collectionName, cancellationToken))
+        await foreach (var dbEntry in this._dbConnector.GetNearestMatchesAsync(this._dbConnection, collectionName, embedding.ToArray(), limit, minRelevanceScore, cancellationToken))
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.DuckDB/DuckDBMemoryStore.cs:155]: I think you can delete `GetAllAsync` now too, right?

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/Database.cs
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.DuckDB/Database.cs:1]: Same here, I think `ReadAllAsync` can be removed.

Comment (lemillermicrosoft) [reply dotnet/src/Connectors/Connectors.Memory.DuckDB/Database.cs:1]: And `DeleteEmptyAsync`

**Review by lemillermicrosoft — APPROVED** [review]
LGTM -- a few methods I think can be removed though. Thanks @colombod!

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/DuckDBMemoryStore.cs (hunk 1 above)
Comment (colombod) [reply dotnet/src/Connectors/Connectors.Memory.DuckDB/DuckDBMemoryStore.cs:155]: let me look!

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/Database.cs
Comment (colombod) [reply dotnet/src/Connectors/Connectors.Memory.DuckDB/Database.cs:1]: done! thanks for pointing that out

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/Database.cs [hunk 2]
```diff
@@ -98,14 +117,23 @@ public async Task CreateCollectionAsync(DuckDBConnection conn, string collection
         }
     }
 
-    public async IAsyncEnumerable<DatabaseEntry> ReadAllAsync(DuckDBConnection conn,
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Security", "CA2100:Review SQL queries for security vulnerabilities", Justification = "Internal method sserializing array of float and numbers")]
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.Memory.DuckDB/Database.cs:120]:
```suggestion
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Security", "CA2100:Review SQL queries for security vulnerabilities", Justification = "Internal method serializing array of float and numbers")]
```

**Review by dmytrostruk — APPROVED** [review]
@colombod Thank you for contribution, having similarity calculation on server side is a huge improvement. 

Idea for future improvements - currently we save all collections and records in one table. I'm wondering if it will be possible to create new table per collection. From DB architecture and data isolation perspective, I think this is the right approach, but would be great to know your opinion on this.

**Comment by colombod** [comment]
> @colombod Thank you for contribution, having similarity calculation on server side is a huge improvement.
> 
> Idea for future improvements - currently we save all collections and records in one table. I'm wondering if it will be possible to create new table per collection. From DB architecture and data isolation perspective, I think this is the right approach, but would be great to know your opinion on this.

Very good point. I kept it this way because, with polyglot notebook, it is possible to send commands to the duckDB and manage persistency and loading from parquet files. From my understanding they deal with a single table.

**Review by lemillermicrosoft — APPROVED** [review]

**Pull request merged**